### PR TITLE
Improved Deepspeed Imports

### LINKED ISSUE #12786: Reduce the import time of pytorch_lightning
## Proposed refactor

The current import time for the pytorch_lightning package on my machine is several seconds. There are some opportunities to improve this.

### Motivation

High import times have an impact on the development and debugging speed. 

### Benchmark

I benchmarked the import time in two environments:
1. Fresh environment with pytorch lightning installed, no extras.
2. My currrent environment, with many extras installed such as loggers, horovod, etc.

To measure the import time, I created a simple file which only imports pytorch_lightning:

```py
import pytorch_lightning as pl
```

Then I use the `importtime` package to measure the time and create a profile:

```
python -X importtime simple.py 2> import.log
```

Finally, I used tuna to visualize the profile:

```
tuna import.log
```

For the fresh environment, the total import time is <2 secs with the following profile:

image

```
>  pip freeze  | grep torch                                            (clean-pl-env) 

pytorch-lightning==1.6.1
torch==1.11.0
torchmetrics==0.7.3
```

For a full development environment, the total import time is >4 seconds:

image

The times vary a bit between multiple runs. However, I have observed that the time is consistently higher when running in an environment with extras installed. Looking at the profiles, it looks like the origin of a large waste of time is coming from our 

`pytorch_lightning.utilities.imports` module where we evaluate some constants at import time:

https://github.com/PyTorchLightning/pytorch-lightning/blob/ae3226ced96e2bc7e62f298d532aaf2290e6ef34/pytorch_lightning/utilities/imports.py#L98-L124

It looks like if a 3rd party package is installed and takes a long time to import, this time gets added to our loading time as well, even if the package never ends up being used. This is because our `_module_available` and `_package_available` implementations attempt to [import the modules](https://github.com/PyTorchLightning/pytorch-lightning/blob/ae3226ced96e2bc7e62f298d532aaf2290e6ef34/pytorch_lightning/utilities/imports.py#L56) to check their availability. This can be very costly. 

### Pitch

Evaluate the import checks lazily. 
Convert 

```py
_X_AVAILALBE = _module_available("x")
```

to

```py
@lru_cache()
def _is_x_available() -> bool:
    return _module_available("x")
```

And investigate other opportunities to improve loading time given the above profile.

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @justusschock @awaelchli @rohitgr7 @borda @akihironitta

### PR DESCRIPTION
**Pull request by Atharva-Phatak**
## What does this PR do?
As mentioned in #12786 PL needed some changes to imports. So as per the dicussion this is my First PR for Deepspeed Imports and later on I will create seprate PRs for Fairscale and all Loggers, etc.

Fixes #12786

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/imports.py [hunk 1]
```diff
@@ -134,9 +134,9 @@ def __repr__(self) -> str:
 
 _APEX_AVAILABLE = _module_available("apex.amp")
 _BAGUA_AVAILABLE = _package_available("bagua")
-_DEEPSPEED_AVAILABLE = _package_available("deepspeed")
-_DEEPSPEED_GREATER_EQUAL_0_5_9 = _DEEPSPEED_AVAILABLE and _compare_version("deepspeed", operator.ge, "0.5.9")
-_DEEPSPEED_GREATER_EQUAL_0_6 = _DEEPSPEED_AVAILABLE and _compare_version("deepspeed", operator.ge, "0.6.0")
+_DEEPSPEED_AVAILABLE = bool(_RequirementAvailable("deepspeed"))
+_DEEPSPEED_GREATER_EQUAL_0_5_9 = _DEEPSPEED_AVAILABLE and bool(_RequirementAvailable("deepspeed>=0.5.9"))
```
Comment (plutasnyy) [inline pytorch_lightning/utilities/imports.py:138]: I wonder if it would not be enough to make it:
`_DEEPSPEED_GREATER_EQUAL_0_5_9 = bool(_RequirementAvailable("deepspeed>=0.5.9"))`

Another thing that bothers me, there was a discussion in the thread about moving imports to places individual use. For example `_DEEPSPEED_GREATER_EQUAL_0_5_9` is used only in https://github.com/PyTorchLightning/pytorch-lightning/blob/90acf909fb0e7bb34aa3a73dda3ea74e705b9230/tests/strategies/test_deepspeed_strategy.py#L49

Maybe we can move this check directly to the file? (same with the others)

Comment (plutasnyy) [inline pytorch_lightning/utilities/imports.py:138]: I wonder if it would not be enough to make it:
`_DEEPSPEED_GREATER_EQUAL_0_5_9 = bool(_RequirementAvailable("deepspeed>=0.5.9"))`

Another thing that bothers me, there was a discussion in the thread about moving imports to places individual use. For example `_DEEPSPEED_GREATER_EQUAL_0_5_9` is used only in https://github.com/PyTorchLightning/pytorch-lightning/blob/90acf909fb0e7bb34aa3a73dda3ea74e705b9230/tests/strategies/test_deepspeed_strategy.py#L49

Maybe we can move this check directly to the file? (same with the others)

Comment (Atharva-Phatak) [reply pytorch_lightning/utilities/imports.py:138]: I agree I will check the test file if it's only required for the test module. Then I will move that accordingly and upated the PR. how does that sound ?

Comment (carmocca) [reply pytorch_lightning/utilities/imports.py]: I helped out by pushing the change to this PR :)

**Comment by plutasnyy** [comment]
Cool, LGTM, thanks @carmocca

**Review by akihironitta — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
